src/agg_bookend_optimize.c

```diff
@@ -0,0 +1,636 @@
+/*
+ * Copyright (c) 2016-2018  Timescale, Inc. All Rights Reserved.
+ *
+ * This file is licensed under the Apache License,
+ * see LICENSE-APACHE at the top level directory.
+ */
+
+/*
+ *	  Optimization for FIRST/LAST aggregate functions.
+ *
+ * This module tries to replace FIRST/LAST aggregate functions by subqueries
+ * of the form
+ *		(SELECT value FROM tab
+ *		 WHERE sort IS NOT NULL AND existing-quals
+ *		 ORDER BY sort ASC/DESC
+ *		 LIMIT 1)
+ * Given a suitable index on sort column, this can be much faster than the
+ * generic scan-all-the-rows aggregation plan.  We can handle multiple
+ * FIRST/LAST aggregates by generating multiple subqueries, and their
+ * orderings can be different.  However, if the query also contains some
+ * other aggregates (eg. MIN/MAX), we will skip optimization since we can't
+ * optimize across different aggregate functions.
+ *
+ *	  Most of the code is borrowed from:
+ *	  src/backend/optimizer/plan/planagg.c
+ *
+ *
+ */
+#include "postgres.h"
+
+#include "access/htup_details.h"
+#include "catalog/pg_aggregate.h"
+#include "catalog/pg_type.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "optimizer/clauses.h"
+#include "optimizer/cost.h"
+#include "optimizer/pathnode.h"
+#include "optimizer/paths.h"
+#include "optimizer/planmain.h"
+#include "optimizer/subselect.h"
+#include "optimizer/tlist.h"
+#include "parser/parsetree.h"
+#include "parser/parse_clause.h"
+#include "rewrite/rewriteManip.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+#include "catalog/pg_proc.h"
+#include <catalog/namespace.h>
+#include "agg_bookend_optimize.h"
+
+typedef struct FirstLastAggInfo
+{
+	MinMaxAggInfo *m_agg_info;	/* reusing MinMaxAggInfo to avoid code
+								 * duplication */
+	Expr	   *sort;			/* Expression to use for ORDER BY */
+} FirstLastAggInfo;
+
+typedef struct MutatorContext
+{
+	MinMaxAggPath *mm_path;
+} MutatorContext;
+
+static const char FIRST[] = "first";
```

Comment:
This is very similar to that case. It's better to have a datastructure that has looked-up OIDS for the functions once instead of looking up function name on  each plan and aggregate. Also oid to oid comparison much cheaper than string comparison.

Reply:
done